gax/backoff_policy.h

```diff
@@ -0,0 +1,181 @@
+// limitations under the License.
+
+#ifndef GOOGLE_GAX_BACKOFF_POLICY_H_
+#define GOOGLE_GAX_BACKOFF_POLICY_H_
+
+#include <algorithm>
+#include <chrono>
+#include <memory>
+#include <random>
+
+#include "googletest/include/gtest/gtest_prod.h"
+
+namespace google {
+namespace gax {
+
+/**
+ * Define the interface for backoff policies.
+ *
+ * The client libraries need to hide partial and temporary failures from the
+ * application. Exponential backoff is generally considered a best practice when
+ * retrying operations. However, the details of how exponetial backoff is
+ * implemented and tuned varies widely. We need to give the users enough
+ * flexibility, and also provide sensible default implementations.
+ *
+ * The client library receives an object of this type and clones a new instance
+ * for each operation. That is, the application provides the library with a
+ * [Prototype](https://en.wikipedia.org/wiki/Prototype_pattern) of the policy
+ * that will be applied to each operation.
+ *
+ * [Truncated Exponential
+ * Backoff](https://cloud.google.com/storage/docs/exponential-backoff) in the
+ * Google Cloud Storage documentation.
+ *
+ */
+class BackoffPolicy {
+ public:
+  virtual ~BackoffPolicy() = default;
+
+  /**
+   * Handle an operation completion.
+   *
+   * This function is called when an operation has failed and needs to be retried.
+   * The decision to retry or not is handled by other policies.
+   *
+   * @return the delay to wait before the next retry attempt.
+   */
+  virtual std::chrono::microseconds OnCompletion() = 0;
+
+  /**
+   * Return a new copy of this object.
+   */
+  virtual std::unique_ptr<BackoffPolicy> clone() const = 0;
+
+  // Used in testing
+ protected:
+  FRIEND_TEST(ExponentialBackoffPolicy, Clone);
+  virtual std::chrono::microseconds CurrentDelay() const = 0;
+  virtual std::chrono::microseconds MaximumDelay() const = 0;
+};
+
+/**
+ * Implements a truncated exponential backoff with randomization policy.
+ *
+ * This policy implements the truncated exponential backoff policy for
+ * retrying operations. After a request fails, and subject to a separate
+ * retry policy, the client library will wait for an initial delay before
+ * trying again. If the second attempt fails the delay time is increased by a factor
+ * of 2. The delay time growth stops at a maximum delay wait time.
+ * The policy also randomizes the delay each time, to avoid
+ * [thundering herd
+ * problem](https://en.wikipedia.org/wiki/Thundering_herd_problem).
+ *
+ * Note: The random number generator used when calculating backoff time in
+ *       OnCompletion is lazily created. Furthermore, GeneratorFactory::Generator
+ *       may touch mutable shared state.
+ *       As a result, OnCompletion is NOT thread-safe.
+ *       All other methods, including clone, are thread-safe.
+ *       ExponentialBackoffPolicy is therefore thread-compatible.
+ */
+template <typename GeneratorFactory>
```

Comment:
Does this mean that a user will need to define a GeneratorFactory type, as you have done in the tests, in order to create an instance of ExponentialBackoffPolicy? That would be very burdensome.

Reply:
We can provide a reference implementation and elide the template from the typename via using, similar to the below using 

```
class MTGenerator {
   static std::unique_ptr<std::mt19937_64> Generator() {
      // use system entropy and other magic to make a well seeded generator
   }
};

using BinaryExponentialBackoffPolicy = ExponentialBackoffPolicy<MTGenerator>;
```

The google-cloud-cpp implementation has an rng factory that lives in `internal/random.h`. Copying that code into a production generator factory class felt outside the scope of the initial BackoffPolicy review.

As a side note, the ClonesHaveDifferentSequences test in google-cloud-cpp has a comment acknowledging that the test could flake if <hell freezes over in June>. This is admittedly extremely unlikely, but "unit tests are deterministic" is a pet peeve of mine but not a hill I'm willing to die on. If you feel that tests don't justify the templating I'll remove it and just add concrete rng seeding.

Reply:
Resolved offline to just make an rng in place without templates.